crates/jstzd/src/image.rs

```diff
@@ -0,0 +1,160 @@
+use async_trait::async_trait;
+use futures_util::stream::StreamExt;
+use std::{collections::HashMap, sync::Arc};
+
+use bollard::{
+    image::{CreateImageOptions, ListImagesOptions},
+    Docker,
+};
+
+#[async_trait]
+pub trait Image {
+    fn image_name(&self) -> &str;
+    fn image_tag(&self) -> &str {
+        "latest"
+    }
+    // Overrides the entrypoint of the image
+    fn entrypoint(&self) -> Option<&str> {
+        None
+    }
+    // If specified, used as a validation on container creation
+    fn exposed_ports(&self) -> &[u16] {
+        &[]
+    }
+    async fn pull_image(&self, client: Arc<Docker>) -> anyhow::Result<()> {
+        if image_exists(self, client.clone()).await {
+            println!("Image: {:?} already exists ", self.image_name());
+            return Ok(());
+        }
+        create_image(self, client.clone()).await
+    }
+}
+
+pub struct DockerImage<'a> {
+    image_name: &'a str,
+    image_tag: Option<&'a str>,
+    entrypoint: Option<&'a str>,
+    exposed_ports: Option<&'a [u16]>,
+}
+
+impl Image for DockerImage<'_> {
+    fn image_name(&self) -> &str {
+        self.image_name
+    }
+
+    fn entrypoint(&self) -> Option<&str> {
+        self.entrypoint
+    }
+
+    fn image_tag(&self) -> &str {
+        self.image_tag.unwrap_or("latest")
+    }
+
+    fn exposed_ports(&self) -> &[u16] {
+        self.exposed_ports.unwrap_or(&[])
+    }
+}
```

Comment:
Suggest alternative name: `GenericImage`, externally this would read as `docker::GenericImage`. 

Additionally remove the lifetimes (all data here can easily be owned)